src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+        {
+            public ArrayBuilder<OrCases>? Set;
+
+            [MemberNotNullWhen(false, nameof(Set))]
+            public bool IsDefault => Set is null;
+
+            internal OrCases StartNewOrCases()
+            {
+                return AddOrSet(new OrCases());
+            }
+
+            internal OrCases AddOrSet(OrCases orCases)
+            {
+                Set ??= ArrayBuilder<OrCases>.GetInstance();
+                Set.Add(orCases);
+                return orCases;
+            }
+
+            public void Dispose()
+            {
+                if (Set is not null)
+                {
+                    foreach (OrCases set in Set)
+                    {
+                        set.Dispose();
+                    }
+
+                    Set.Free();
+                }
+            }
+
+#if DEBUG
+            public override string ToString()
+            {
+                if (IsDefault)
+                    return "DEFAULT";
+
+                var builder = new StringBuilder();
+                foreach (var set in Set)
+                {
+                    builder.AppendLine("Set:");
+                    foreach (var @case in set.Cases)
+                    {
+                        builder.Append(@case.pattern.DumpSource());
+                        if (@case.syntax is { } syntax)
+                        {
+                            builder.Append("  => ");
+                            builder.Append(syntax.ToString());
+                        }
+
+                        builder.AppendLine();
+                    }
+
+                    builder.AppendLine();
+                }
+
+                return builder.ToString();
+            }
+#endif
+        }
+
+        /// <summary>
+        /// When we have a single sequence of `or` patterns, such as `A or B or C`
+        /// we can expand it to separate cases: `A`, `B` and `C`.
+        /// This composes. So a nested `or` sequence can also be expanded: `X and (A or B or C)`
+        /// can be expanded to `X and A`, `X and B` and `X and C`.
+        /// </summary>
+        private struct OrCases : IDisposable
+        {
+            public ArrayBuilder<(BoundPattern pattern, SyntaxNode? syntax)> Cases;
+
+            public OrCases()
+            {
+                Cases = ArrayBuilder<(BoundPattern pattern, SyntaxNode? syntax)>.GetInstance();
+            }
+
+            public void Add(BoundPattern pattern, SyntaxNode? syntax)
+            {
+                Cases.Add((pattern, pattern.WasCompilerGenerated ? null : syntax));
+            }
+
+            public void Dispose()
+            {
+                Cases.Free();
+            }
+        }
+
+        // The purpose of this rewriter is to push `not` patterns down the pattern tree.
+        // It needs to expand composite patterns in the process.
+        // TODO2 some patterns get erased (variable declarations?)
+        private class MoveNotPatternsDownRewriter : BoundTreeRewriterWithStackGuard
```

Comment:
>MoveNotPatternsDownRewriter [](http://example.com/codeflow?start=22&length=27)

It would be good add some kind of assert to catch a situation when this visitor fails to override one of the ```Visit*Pattern``` methods